velox/core/PlanNode.cpp

```diff
@@ -846,35 +846,90 @@ PlanNodePtr MergeJoinNode::create(const folly::dynamic& obj, void* context) {
       outputType);
 }
 
-CrossJoinNode::CrossJoinNode(
+NestedLoopJoinNode::NestedLoopJoinNode(
     const PlanNodeId& id,
+    JoinType joinType,
+    TypedExprPtr joinCondition,
     PlanNodePtr left,
     PlanNodePtr right,
-    RowTypePtr outputType)
+    const RowTypePtr& outputType)
     : PlanNode(id),
+      joinType_(joinType),
+      joinCondition_(std::move(joinCondition)),
       sources_({std::move(left), std::move(right)}),
-      outputType_(std::move(outputType)) {}
+      outputType_(outputType) {
+  VELOX_CHECK(
+      core::isInnerJoin(joinType_) || core::isLeftJoin(joinType_) ||
+          core::isRightJoin(joinType_) || core::isFullJoin(joinType_),
+      "Currently NestedLoopJoin only supports inner and outer join");
 
-void CrossJoinNode::addDetails(std::stringstream& /* stream */) const {
-  // Nothing to add.
+  auto leftType = sources_[0]->outputType();
+  auto rightType = sources_[1]->outputType();
+  for (const auto& name : outputType_->names()) {
+    const bool leftContains = leftType->containsChild(name);
+    const bool rightContains = rightType->containsChild(name);
+    VELOX_CHECK(
+        !(leftContains && rightContains),
+        "Duplicate column name found on join's left and right sides: {}",
+        name);
+    VELOX_CHECK(
+        leftContains || rightContains,
+        "Join's output column not found in either left or right sides: {}",
+        name);
+  }
+}
+
+NestedLoopJoinNode::NestedLoopJoinNode(
+    const PlanNodeId& id,
+    PlanNodePtr left,
+    PlanNodePtr right,
+    const RowTypePtr& outputType)
+    : NestedLoopJoinNode(
+          id,
+          JoinType::kInner,
+          nullptr,
+          left,
+          right,
+          outputType) {}
+
+void NestedLoopJoinNode::addDetails(std::stringstream& stream) const {
+  stream << joinTypeName(joinType_);
+  if (joinCondition_) {
+    stream << ", joinCondition: " << joinCondition_->toString();
+  }
 }
 
-folly::dynamic CrossJoinNode::serialize() const {
+folly::dynamic NestedLoopJoinNode::serialize() const {
   auto obj = PlanNode::serialize();
+  obj["joinType"] = joinTypeName(joinType_);
+  if (joinCondition_) {
+    obj["joinCondition"] = joinCondition_->serialize();
+  }
   obj["outputType"] = outputType_->serialize();
   return obj;
 }
 
-// static
-PlanNodePtr CrossJoinNode::create(const folly::dynamic& obj, void* context) {
+PlanNodePtr NestedLoopJoinNode::create(
+    const folly::dynamic& obj,
+    void* context) {
   auto sources = deserializeSources(obj, context);
   VELOX_CHECK_EQ(2, sources.size());
 
-  return std::make_shared<CrossJoinNode>(
+  TypedExprPtr joinCondition;
+  if (obj.count("joinCondition")) {
+    joinCondition =
+        ISerializable::deserialize<ITypedExpr>(obj["joinCondition"]);
```

Comment:
Deserialization of ITypedExpr may need access to memory pool, hence, need to pass 'context'.

Reply:
Several invocations of `deserialize<ITypedExpr>` in `PlanNode.cpp` do not pass `context`, shoud we fix them?

https://github.com/facebookincubator/velox/blob/3e6e7ecef3a1129164738515b19793b0b494c319/velox/core/PlanNode.cpp#L802

https://github.com/facebookincubator/velox/blob/3e6e7ecef3a1129164738515b19793b0b494c319/velox/core/PlanNode.cpp#L833

https://github.com/facebookincubator/velox/blob/3e6e7ecef3a1129164738515b19793b0b494c319/velox/core/PlanNode.cpp#L1057

https://github.com/facebookincubator/velox/blob/3e6e7ecef3a1129164738515b19793b0b494c319/velox/core/PlanNode.cpp#L1062

https://github.com/facebookincubator/velox/blob/3e6e7ecef3a1129164738515b19793b0b494c319/velox/core/PlanNode.cpp#L1575

Reply:
Yes, please.